zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -2788,40 +2798,106 @@ public void setACL(final String path, List<ACL> acl, int version, StatCallback c
+        }
+        updateNextPage(outputNextPage, response.getNextPageCzxid(), response.getNextPageCzxidOffset());
+
         return response.getChildren();
     }
 
+    /**
+     * Returns a list of all the children given the path.
+     * <p>
+     * The difference between this API and {@link #getChildren(String, boolean)} is:
+     * when there are lots of children and the network buffer exceeds {@code jute.maxbuffer},
+     * this API will fetch the children using pagination and be able to return all children;
+     * while {@link #getChildren(String, boolean)} will fail.
+     * <p>
+     * The final list of children returned is NOT strongly consistent with the server's data:
+     * the list might contain some deleted children if some children are deleted before the last page is fetched.
+     * <p>
+     * If the watch is true and the call is successful (no exception is thrown),
+     # a watch will be left on the node with the given path. The watch will be
+     * triggered by a successful operation that deletes the node of the given
+     * path or creates/deletes a child under the node.
+     *
+     * @param path the path of the parent node
+     * @param watch whether or not leave a watch on the given node
+     * @return a list of all children of the given path
+     * @throws KeeperException if the server signals an error with a non-zero error code.
+     * @throws InterruptedException if the server transaction is interrupted.
+     */
+    public List<String> getAllChildrenPaginated(String path, boolean watch)
+            throws KeeperException, InterruptedException {
+        PaginationNextPage nextPage = new PaginationNextPage();
+        Watcher watcher = watch ? watchManager.defaultWatcher : null;
+        List<String> childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, 0L, 0, null, nextPage);
+        if (nextPage.getMinCzxid() == ZooDefs.GetChildrenPaginated.lastPageMinCzxid) {
+            return childrenPaginated;
+        }
+
+        Set<String> childrenSet = new HashSet<>(childrenPaginated);
+        do {
+            long minCzxid = nextPage.getMinCzxid();
+            // If a child with the same czxid is returned in the previous page, and then deleted
+            // on the server, the children not in the previous page but offset is less than czxidOffset
+            // would be missed in the next page. Use the last child in the previous page to determine whether or not
+            // the deletion case happens. If it happens, retry fetching the page starting from offset 0.
+            int czxidOffset = nextPage.getMinCzxidOffset() - 1;
+            childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, minCzxid, czxidOffset, null, nextPage);
+            if (!childrenPaginated.isEmpty() && !childrenSet.contains(childrenPaginated.get(0))) {
+                childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, minCzxid, 0, null, nextPage);
```

Comment:
We dont provide data consistent, right? Then why we need to provide retry for consistent check?

Reply:
Replied in the above thread: https://github.com/linkedin/zookeeper/pull/8/files/c400bc183a7c61f35a524a57f828715d103badf9#r532831492

Reply:
Add a log for this retry.

Reply:
Add a log for this retry for now would be good idea. Later, we may later consider add native zookeeper mbean.